processor/signoztailsampler/evaluator.go

```diff
@@ -0,0 +1,141 @@
+	// should be applied
+	filters := prepareFilterEvaluators(logger, policyCfg.PolicyFilterCfg)
+	// todo(amol): need to handle situations with zero filters
+
+	// list of sub-policies evaluators
+	subpolicies := make([]sampling.PolicyEvaluator, len(policyCfg.SubPolicies))
+	for _, subRule := range policyCfg.SubPolicies {
+		subPolicy := NewDefaultEvaluator(logger, &subRule)
+		subpolicies = append(subpolicies, subPolicy)
+	}
+
+	// sampling is applied only when filter conditions are met
+	var sampler sampling.PolicyEvaluator
+	var samplingMethod string
+	switch policyCfg.SamplingPercentage {
+	case 0:
+		samplingMethod = "exclude_all"
+		sampler = sampling.NewAlwaysUnsample(logger)
+	case 100:
+		samplingMethod = "include_all"
+		sampler = sampling.NewAlwaysSample(logger)
+	default:
+		samplingMethod = "probabilistic"
+		sampler = sampling.NewProbabilisticSampler(logger, "21d2", policyCfg.SamplingPercentage)
+	}
+
+	return &defaultEvaluator{
+		name:           policyCfg.Name,
+		root:           policyCfg.Root,
+		sampler:        sampler,
+		samplingMethod: samplingMethod,
+		filterOperator: filterOperator,
+		filters:        filters,
+		subpolicies:    subpolicies,
+	}
+}
+
+func prepareFilterEvaluators(logger *zap.Logger, policyFilterCfg PolicyFilterCfg) []sampling.PolicyEvaluator {
+	var filterEvaluators []sampling.PolicyEvaluator
+	for _, s := range policyFilterCfg.StringAttributeCfgs {
+		filterEvaluators = append(filterEvaluators, sampling.NewStringAttributeFilter(logger, s.Key, s.Values, s.EnabledRegexMatching, s.CacheMaxSize, s.InvertMatch))
+	}
+
+	for _, n := range policyFilterCfg.NumericAttributeCfgs {
+		filterEvaluators = append(filterEvaluators, sampling.NewNumericAttributeFilter(logger, n.Key, n.MinValue, n.MaxValue))
+	}
+	// todo: status filter
+	return filterEvaluators
+}
+
+func (de *defaultEvaluator) Evaluate(traceId pcommon.TraceID, trace *sampling.TraceData) (sampling.Decision, error) {
+
+	// todo: explore doing this in parallel
+	for _, sp := range de.subpolicies {
+		decision, err := sp.Evaluate(traceId, trace)
+		if err != nil {
+			// todo: consider adding health for each evaluator
+			// to avoid printing log messages for each trace
+			zap.S().Errorf("failed to evaluate trace:", traceId)
+			continue
+		}
+
+		if decision != sampling.NoResult {
+			// found a result, exit
+			return decision, nil
+		}
+	}
+
+	// filterMatched captures when atleast one filter matches
+	var filterMatched bool
+
+	for _, fe := range de.filters {
+
+		filterResult, err := fe.Evaluate(traceId, trace)
+		if err != nil {
+			return sampling.Error, nil
+		}
+		// since filters can also be sampling evaluators,
+		// here we consider "sampled = filter matched" and
+		// "no sampled == filter did not match"
+
+		if filterResult == sampling.Sampled {
+			filterMatched = true
+		}
+
+		if filterResult == sampling.NotSampled {
+			if de.filterOperator == "AND" {
+				// filter condition failed, return no op
+				return sampling.NoResult, nil
+			}
```

Comment:
Why specifically `AND`? I don't how `OR` will have a different result.

Reply:
In case of OR, if filterResult is set even once, the outcome will be TRUE. 
in case of AND, we dont want even a single negative result.